File: js/11066.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map((e) => e.split(' ').map(Number));

let idx = 0;
const getLine = () => inputs[idx++];

let [T] = getLine();

while (T--) {
  const [K] = getLine();
  const files = getLine();

  // dp[i][j] i부터 j까지 파일을 합쳤을 때 최소비용
  const dp = Array(K + 1)
    .fill()
    .map(() => Array(K + 1).fill(0));
  const sums = Array(K + 1).fill(0);

  for (let i = 0; i < K; i++) {
    sums[i + 1] = sums[i] + files[i];
  }
  for (let gap = 1; gap < K; gap++) {
    for (let start = 1; start <= K - gap; start++) {
      const end = start + gap;
      dp[start][end] = Infinity;
      for (let k = start; k < end; k++) {
        dp[start][end] = Math.min(
          dp[start][end],
          dp[start][k] + dp[k + 1][end] + sums[end] - sums[start - 1]
        );
      }
    }
  }
  console.log(dp);

  console.log(dp[1][K]);
}

File: c++/11066.cpp
#include <algorithm>
#include <iostream>

#define INF 1e9

using namespace std;

int main() {
  freopen("example.txt", "r", stdin);
  int T;
  cin >> T;

  while (T--) {
    int K;
    cin >> K;

    int sums[501] = {0};
    int dp[501][501] = {0};

    for (int i = 0; i < K; i++) {
      int a;
      cin >> a;
      sums[i + 1] = sums[i] + a;
    }

    for (int gap = 1; gap < K; gap++) {
      for (int start = 1; start <= K - gap; start++) {
        int end = start + gap;
        dp[start][end] = INF;
        for (int k = start; k < end; k++) {
          dp[start][end] = min(dp[start][end], dp[start][k] + dp[k + 1][end] +
                                                   sums[end] - sums[start - 1]);
        }
      }
    }

    cout << dp[1][K] << '\n';
  }

  return 0;
}